File: leetCode/js/implement_trie(prefix_tree).js
// 208. Implement Trie(Prefix Tree)
// Medium

// Implement a trie with insert, search, and startsWith methods.

//     Example:

// Trie trie = new Trie();

// trie.insert("apple");
// trie.search("apple");   // returns true
// trie.search("app");     // returns false
// trie.startsWith("app"); // returns true
// trie.insert("app");
// trie.search("app");     // returns true

// Note:

// You may assume that all inputs are consist of lowercase letters a - z.
// All inputs are guaranteed to be non - empty strings.

class Node {
    constructor() {
        this.children = {};
        this.isTerminal = false;
    }
}

class Trie {
    constructor() {
        this.root = new Node();
    }

    insert(word, root = this.root) {
        let letter = word[0];

        if (!(letter in root.children)) {
            root.children[letter] = new Node();
        }

        if (word.length === 1) {
            root.children[letter].isTerminal = true;
        } else {
            this.insert(word.slice(1), root.children[letter]);
        }
    };

    search(word, root = this.root) {
        let letter = word[0];
        if (word.length === 0) {
            if (root.isTerminal) {
                return true;
            } else {
                return false;
            }
        };

        if (letter in root.children) {
            return this.search(word.slice(1), root.children[letter]);
        } else {
            return false;
        }
    };

    startsWith(word, root = this.root) {
        let letter = word[0];
        if (word.length === 0) return true;

        if (letter in root.children) {
            return this.startsWith(word.slice(1), root.children[letter]);
        } else {
            return false;
        }
    };
};

File: leetCode/py3/implement_trie(prefix_tree).py
# 208. Implement Trie (Prefix Tree)
# Medium

# Implement a trie with insert, search, and startsWith methods.

# Example:

# Trie trie = new Trie();

# trie.insert("apple");
# trie.search("apple");   // returns true
# trie.search("app");     // returns false
# trie.startsWith("app"); // returns true
# trie.insert("app");   
# trie.search("app");     // returns true

# Note:

#     You may assume that all inputs are consist of lowercase letters a-z.
#     All inputs are guaranteed to be non-empty strings.

class Node:

    def __init__(self):
        self.children = {};
        self.is_terminal = False;

class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = Node()

    def insert(self, word: str, root=None) -> None:
        """
        Inserts a word into the trie.
        """
        root = root or self.root
        letter = word[0]
        if not root.children.get(letter, False):
            root.children[letter] = Node()
        if len(word) == 1:
            root.children[letter].is_terminal = True
        else:
            self.insert(word[1:], root.children[letter])

    def search(self, word: str, root=None) -> bool:
        """
        Returns if the word is in the trie.
        """
        root = root or self.root
        if word == '':
            return True if root.is_terminal else False
        letter = word[0]
        if root.children.get(letter, False):
            return self.search(word[1:], root.children[letter])
        else:
            return False

    def startsWith(self, prefix: str, root=None) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        root = root or self.root
        if prefix == '': return True
        letter = prefix[0]
        if root.children.get(letter, False):
            return self.startsWith(prefix[1:], root.children[letter])
        else:
            return False
